src/natpmp.cpp

```diff
@@ -278,7 +277,8 @@ void natpmp::try_next_mapping(int const i)
 
 	auto const m = std::find_if(
 		m_mappings.begin(), m_mappings.end()
-		, [] (mapping_t const& ma) { return ma.act != mapping_t::action::none; });
+		, [] (mapping_t const& ma) { return ma.act != mapping_t::action::none &&
```

Comment:
i can not reproduce stack overflow  :-(

int natpmp::add_mapping(portmap_protocol const p, int const external_port
	, tcp::endpoint const local_ep)
{
#ifdef _DEBUG
	static bool is_first = false;
	if (!is_first)
	{
		is_first = true;
		m_mappings.resize(4);

		m_mappings[0].protocol = portmap_protocol::none;
		m_mappings[1].protocol = portmap_protocol::none;
		m_mappings[2].protocol = portmap_protocol::tcp;
		m_mappings[3].protocol = portmap_protocol::udp;

		m_mappings[0].act = mapping_t::action::add;
		m_mappings[1].act = mapping_t::action::none;
		m_mappings[2].act = mapping_t::action::none;
		m_mappings[3].act = mapping_t::action::none;

		m_mappings[0].external_port = 8999;
		m_mappings[1].external_port = 8999;
		m_mappings[2].external_port = 8999;
		m_mappings[3].external_port = 8999;

		m_mappings[0].local_port = 8999;
		m_mappings[1].local_port = 8999;
		m_mappings[2].local_port = 8999;
		m_mappings[3].local_port = 8999;

		m_mappings[0].map_sent = true;
		m_mappings[1].map_sent = true;
		m_mappings[2].map_sent = true;
		m_mappings[3].map_sent = true;

		m_mappings[0].outstanding_request = false;
		m_mappings[1].outstanding_request = false;
		m_mappings[2].outstanding_request = false;
		m_mappings[3].outstanding_request = false;

		m_mappings[0].expires = aux::time_now();
		m_mappings[1].expires = aux::time_now();
		m_mappings[2].expires = aux::time_now();
		m_mappings[3].expires = aux::time_now();
	}